### ISSUE TITLE
[Bug] unhandledRejection (Target.attachToTarget)

### ISSUE THREAD

**Issue by wintrdotcom** [berstend/puppeteer-extra#470]
**Describe the bug**

Hey there, reproducing my code snippet will lead to an `unhandledRejection`.

It basically opens a Browser instance and closes it before creating a new page, therefore, it fails as expected. The problem is that the error is not correctly catched.

If I reproduce the same code snippet commenting `puppeteer.use(pluginStealth())`, the error will be correctly catched so I guess the problem comes from `puppeteer-extra` rather than from `puppeteer` itself.

**Code Snippet**

```javascript
const puppeteer = require('puppeteer-extra')
const pluginStealth = require('puppeteer-extra-plugin-stealth')

puppeteer.use(pluginStealth())

puppeteer.launch({
	headless: false,
	args: [
		'--window-position=0,0',
		'--disable-infobars',
		'--disable-extensions',
		'--disable-accelerated-2d-canvas',
		'--disable-gpu',
		'--window-size=1920x1080',
		'--disable-dev-shm-usage',
		'--disable-setuid-sandbox',
		'--no-first-run',
		'--no-sandbox',
		'--no-zygote',
		'--single-process',
		'--deterministic-fetch',
		'--enable-features=NetworkService',
		'--ignore-certificate-errors'
	]
})
.then((browser) => {
    browser.close()

	return browser.newPage()
})
.catch((err) => {
    console.log(err)
})
```

**What is happening**

```javascript
(node:18424) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:18424) UnhandledPromiseRejectionWarning: Error: Protocol error (Target.attachToTarget): No target with given id found
    at C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:71:63
    at new Promise (<anonymous>)
    at Connection.send (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:70:16)
    at Connection.createSession (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:141:42)
    at Target._sessionFactory (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Browser.js:178:91)
    at Target.page (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Target.js:72:38)
    at Plugin._onTargetCreated (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer-extra-plugin\dist\index.cjs.js:489:39)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
(node:18424) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 2)
```

**What should happen**

```javascript
Error: Protocol error (Target.attachToTarget): No target with given id found
    at C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:71:63
    at new Promise (<anonymous>)
    at Connection.send (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:70:16)
    at Connection.createSession (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Connection.js:141:42)
    at Target._sessionFactory (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Browser.js:178:91)
    at Target.page (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer\lib\cjs\puppeteer\common\Target.js:72:38)
    at Plugin._onTargetCreated (C:\Users\xxxx\Desktop\lol\node_modules\puppeteer-extra-plugin\dist\index.cjs.js:489:39)
    at processTicksAndRejections (internal/process/task_queues.js:93:5) {
  message: 'Protocol error (Target.attachToTarget): No target with given id found'
}
```

**Versions**

  System:
    OS: Windows 10 10.0.19042
    CPU: (6) x64 Intel(R) Core(TM) i5-8400 CPU @ 2.80GHz
    Memory: 4.24 GB / 7.92 GB
  Binaries:
    Node: 12.13.1 - C:\Program Files\nodejs\node.EXE
    npm: 6.12.1 - C:\Program Files\nodejs\npm.CMD

**Comment by prescience-data** [comment]
have you tried:

```typescript

;(async () => {
  // Launch a browser...
  const browser = puppeteer.launch({
    headless: false,
    args: [
      "--window-position=0,0",
      "--disable-infobars",
      "--disable-extensions",
      "--disable-accelerated-2d-canvas",
      "--disable-gpu",
      "--window-size=1920x1080",
      "--disable-dev-shm-usage",
      "--disable-setuid-sandbox",
      "--no-first-run",
      "--no-sandbox",
      "--no-zygote",
      "--single-process",
      "--deterministic-fetch",
      "--enable-features=NetworkService",
      "--ignore-certificate-errors"
    ]
  })
  // Create a new page...
  const page = await browser.newPage()

  // Do stuff...
  
  await page.goto("https://example.com")
})()
```

**Comment by wintrdotcom** [comment]
Well, if you are not closing the browser, the bug won’t happen, maybe I haven’t been clear enough.

The browser instance can be closed for many reasons, crash, server outage, etc. What I noticed is that on browser instance close (that I trigger myself to show you the problem), the errors are not correctly catched because there is some uncatched promises somewhere in the code.

Therefore, there is no way to correctly handle those errors.

**Comment by prescience-data** [comment]
Maybe wrap the whole thing in a try catch block?

**Comment by wintrdotcom** [comment]
It has the exact same effect, the unhandled promise will always be logged. The correct way to fix that would be to correctly catch the called promises.

**Comment by prescience-data** [comment]
Have you had a quick look through the source to find where the disparity might be coming from?

**Comment by berstend** [comment]
To quote myself here from our discord channel:

> hmm, that's not necessarily a fatal error - it's more of an FYI that the Target closed while still being worked with
that can happen if a page is navigated/closed immediately after loading
as for why that happens only with the stealth plugin enabled (presumably): The plugin will do a bunch of things with a page, so the chance of this happening is enlarged
look into handleUncaughtExceptions handlers if you want to suppress these messages (in case they don't affect your business logic)

> keep in mind that it's best practice to selectively check for errors you want to supress and rethrow the rest in these global handlers: https://nodejs.org/api/process.html#process_event_uncaughtexception

Another user suggest to explicitly close the page before closing the browser to mitigate this issue.

Yet another user mentioned launch with a `'--disable-dev-shm-usage'` flag improve this behavior.

**TL;DR: This is not really a bug but relatively normal behavior.**

If you want to suppress those errors (or make them non-fatal) use a global catch handler (make sure to selectively mute errors there though):
```js
// untested but something along these lines should work
process.on('uncaughtException', (err, origin) => {
  const isTargetClosedError = err?.message.includes('Protocol error (Target.attachToTarget): No target with given id found')
  if (!isTargetClosedError) { throw err } // Re-throw all other errors
});
```

See here for how common that issue is: https://stackoverflow.com/a/51989560

Closing this unless there's clear/specific evidence that extra or stealth has a bug in this regard that needs fixing.

**Comment by berstend** [comment]
@wintrdotcom Wait, I just had a closer look at your code:
```js
.then((browser) => {
    browser.close()

   return browser.newPage()
})
```
This is just broken code? you're closing the browser and then working on it - that this even works in vanilla pptr is just by chance. 😄

**Comment by clouedoc** [comment]
I stumbled upon this problem today. I usually do a lot of debugging to find out where the problem comes from, but I thought that a `unhandledRejection` handler could the trick for this case.

Here is my code, for reference:

```typescript
export class UnhandledRejectionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnhandledRejectionError";
  }
}

process.on("unhandledRejection", (reason, promise) => {
  const isTargetClosedError = `${reason}`.includes(
    "Session closed. Most likely the page has been closed."
  );

  const metadata = {
    reason,
    promise,
  };

  if (isTargetClosedError) {
    // log is a Winston Logger. You can safely replace it with console.log or console.warn
    log.warn("Target closed error", metadata);
    return;
  }

  throw new UnhandledRejectionError(
    "Unhandled Promise rejection. Infos: " + JSON.stringify(metadata)
  );
});
```

**Comment by berstend** [comment]
@clouedoc so is the code you've posted working for you? :-)

**Comment by clouedoc** [comment]
> @clouedoc so is the code you've posted working for you? :-)

I tested the code and I can confirm that it does the following:

1. Catches the unhandled session/target errors
2. Warns me about them without crashing the process
3. Still throws other unhandled promise rejections

I would say that yes, it works. Feel free to point it out if you've found something odd in it 😉!

**Comment by wintrdotcom** [comment]
Hey again, thanks for this awesome answers, I already tested a similar solution using process.on("unhandledRejection"... previously

It seems that it is catching some of the unhandledRejections but still throws a node error sometimes, I'm going to give it another shot using @clouedoc solution.

I'm running a 5m+ various headless requests a day so I have enough data to sort this out quickly, I'll give you an update on this in a couple of days.

Thanks again!

**Comment by clouedoc** [comment]
My previous code wasn't logging the full exception.

@wintrdotcom you may want to update your code

I now use: 

```typescript
import delay from "delay";
import { log } from "./log";

export class UnhandledRejectionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "UnhandledRejectionError";
  }
}

export function registerUncaughtListeners() {
  // use this to report fatal errors to GCP
  process.on("uncaughtException", async (err) => {
    // [...]
  });

  process.on("unhandledRejection", (reason: any, promise) => {
    const isSessionClosedError = `${reason}`.includes(
      "Session closed. Most likely the page has been closed."
    );
    const isTargetClosedError = `${reason}`.includes("Target closed.");

    const metadata = {
      // CHANGE: use reason.stack when possible to get the full stacktrace
      reason: reason?.stack || reason,
      promise,
    };

    if (isTargetClosedError) {
      log.warn("Target closed error", metadata);
      return;
    } else if (isSessionClosedError) {
      log.warn("Session closed error", metadata);
      return;
    }

    const err = new UnhandledRejectionError(
      "Unhandled Promise rejection. Infos: " + JSON.stringify(metadata)
    );
    log.error("fatal uncaught exception error: " + err, {
      name: err.name,
      metadata,
    });
    throw err;
  });
}
```

Don't forget to call this function to attach the listeners.

```typescript
// place this after having exported your global logging instance
registerUncaughtListeners();
```

**Comment by wintrdotcom** [comment]
After using latest @clouedoc answer with `registerUncaughtListeners` function, I still have some uncatched UnhandledRejectionErrors popping from time to time, the most incredible is that I even tried to put:

`
process.on('uncaughtException', (err) => {
    	return
})
    
process.on('unhandledRejection', (reason, promise) => {
        return
 })
`

in the very begining of my app on a test server and it still reports some errors after a large amount of requests :x

I don't really know how it is possible.

**Comment by clouedoc** [comment]
I believe that [you should exit on `uncaughtException`](https://nodejs.org/api/process.html#process_warning_using_uncaughtexception_correctly).

Are the errors you receive uncaught exceptions or uncaught rejections?

**Comment by wintrdotcom** [comment]
They are `unhandledRejection`:

```
You have triggered an unhandledRejection, you may have forgotten to catch a Promise rejection:
Error: Protocol error (Target.attachToTarget): No target with given id found
    at /var/www/node_modules/puppeteer/lib/Connection.js:74:56
    at new Promise (<anonymous>)
    at Connection.send (/var/www/node_modules/puppeteer/lib/Connection.js:73:12)
    at Connection.createSession (/var/www/node_modules/puppeteer/lib/Connection.js:153:36)
    at Target._sessionFactory (/var/www/node_modules/puppeteer/lib/Browser.js:116:75)
    at Target.worker (/var/www/node_modules/puppeteer/lib/Target.js:87:34)
    at Target.<anonymous> (/var/www/node_modules/puppeteer/lib/helper.js:112:23)
    at Plugin._onTargetCreated (/var/www/node_modules/puppeteer-extra-plugin/dist/index.cjs.js:498:41)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:97:5)
```

**Comment by clouedoc** [comment]
I really don't know what could be the issue... Maybe an `await` that you forgot which leads to the browser being closed?

Did you manage to fix your issue 🤔?

**Comment by wintrdotcom** [comment]
I'm not using async await but exclusively promises, I did not manage to fix the issues yet, the stack trace still show that the error comes from /puppeteer/lib/Connection.js that is called by /puppeteer-extra-plugin/dist/index.cjs.js so I'm almost sure it does not come from my code.

My real interrogation is how the heck is it possible to still have those warnings after putting:

```
process.on('uncaughtException', (err) => {
    	return
})
    
process.on('unhandledRejection', (reason, promise) => {
        return
 })
```

In the very begining of my app.js in a basic express model/route/controller app ^^

Thanks for the support anyways, I will keep you in touch once I realize why.

**Comment by clouedoc** [comment]
@wintrdotcom 

I just stumbled upon another unhandled promise problem today.
I've found a way to catch them, actually!

![image](https://user-images.githubusercontent.com/13921610/115158161-2f571080-a08d-11eb-9f01-16fc630d1771.png)


**Catched:**
![image](https://user-images.githubusercontent.com/13921610/115158142-13ec0580-a08d-11eb-982a-d434159844da.png)
**Fixed:**
![image](https://user-images.githubusercontent.com/13921610/115158183-4ac21b80-a08d-11eb-82e0-087314b0254b.png)

It also catches when I forget to `await` (and which could throw `UnhandledPromiseRejections`

![image](https://user-images.githubusercontent.com/13921610/115158108-e1daa380-a08c-11eb-829a-c1a12d7697e0.png)

This is all done thanks to the `@typescript-eslint/no-floating-promises` ESLint rule.
If it doesn't work (e.g. your project is good old JavaScript), there's also an ESLint plugin for that.

All you need can be found here: https://stackoverflow.com/questions/63069344/can-eslint-help-you-prevent-unhandled-promise-rejections

My `.eslintrc.js` for reference:

```js
module.exports = {
  env: {
    es2021: true,
    node: true,
  },
  extends: ["eslint:recommended", "plugin:@typescript-eslint/recommended"],
  parser: "@typescript-eslint/parser",
  parserOptions: {
    ecmaVersion: 12,
    sourceType: "module",
    project: "tsconfig.json",
  },
  plugins: ["@typescript-eslint"],
  rules: {
    "@typescript-eslint/no-floating-promises": "error",
  },
};
```

I highly suggest that you try this! This might solve the root of your problem.

**Comment by wintrdotcom** [comment]
Hey there, I ran many tests and checked for uncatched promises as you previously advised, my code was perfect and I didn't find anything wrong in it. However, I can't manage to reproduce the bug using Puppeteer 8.0 so I guess it has been fixed somehow.

Anyways, thanks for your help and sorry for the late answer, I wanted to be sure to not populate this issue with too many answers without having everything tested.

If someone struggles on it, make sure you're using the latest version of Puppeteer

**Comment by berstend** [comment]
This should now be fixed through #555:
```
Successfully published:
 - puppeteer-extra-plugin-adblocker@2.12.0
 - puppeteer-extra-plugin-anonymize-ua@2.3.0
 - puppeteer-extra-plugin-block-resources@2.3.0
 - puppeteer-extra-plugin-click-and-wait@2.3.0
 - puppeteer-extra-plugin-devtools@2.3.0
 - puppeteer-extra-plugin-flash@2.3.0
 - puppeteer-extra-plugin-font-size@2.3.0
 - puppeteer-extra-plugin-recaptcha@3.5.0
 - puppeteer-extra-plugin-repl@2.3.0
 - puppeteer-extra-plugin-stealth@2.8.0
 - puppeteer-extra-plugin-user-data-dir@2.3.0
 - puppeteer-extra-plugin-user-preferences@2.3.0
 - puppeteer-extra-plugin@3.2.0
 - puppeteer-extra@3.2.0
```
